noir-projects/noir-protocol-circuits/crates/types/src/merkle_tree/variable_merkle_tree.nr

```diff
@@ -0,0 +1,185 @@
+use crate::{traits::Empty, hash::accumulate_sha256};
+global PWR_LIMIT = 8;
+// N = maximum leaves
+// For now we only care about the root
+struct VariableMerkleTree<N> {
+    leaves: [Field; N],
+    root: Field,
+}
+unconstrained fn get_surrounding_pwrs_2(input: u32) -> (u32, u32, u8) {
+    let mut next_power_2 = 0;
+    let mut prev_power_2 = 0;
+    let mut height = 0;
+    let mut stop = false;
+    for i in 0..PWR_LIMIT {
+        if !stop {
+            height = i;
+            prev_power_2 = next_power_2;
+            next_power_2 = 2 << i;
+            if input <= next_power_2 {
+                stop = true;
+            }
+        }
+    }
+    (prev_power_2, next_power_2, height)
+}
+
+// This calculates the root of the minimal size merkle tree required
+// to store num_non_empty_leaves
+// Since we cannot isolate branches, it doesn't cost fewer gates than using
+// MerkleTree on the full array of elements N, but is slightly cheaper on-chain
+// and cleaner elsewhere.
+impl<N> VariableMerkleTree<N> {
+    pub fn new_sha(leaves: [Field; N], num_non_empty_leaves: u32) -> Self {
+        // Find size of tree required
+        let (prev_power_2, next_power_2, height) = get_surrounding_pwrs_2(num_non_empty_leaves);
+        // If we have no num_non_empty_leaves, we use a 'tree' of 2 leaves
+        // TODO perhaps we just return 0 in this case?
+        assert((num_non_empty_leaves == 0) | (num_non_empty_leaves > prev_power_2));
+        assert(num_non_empty_leaves <= next_power_2);
+        assert(next_power_2 == 2 << height);
+        // hash base layer
+        let mut stop = false;
+        let mut nodes = [0; N];
+        let mut base_layer_end = next_power_2 / 2;
```

Comment:
Can we use `prev_power_2`? I don't think it's used and checked (== 2 << (height - 1)).